build-coordinator/src/main/java/org/jboss/pnc/coordinator/builder/BuildCoordinator.java

```diff
@@ -203,113 +157,149 @@ private boolean rejectAlreadySubmitted(BuildTask buildTask) {
+                    coordinationStatus = BuildCoordinationStatus.DONE_WITH_ERRORS;
+                } else {
+                    throw new BuildCoordinationException("Failed task should have set exception or failed reason status.");
+                }
+            } else {
+                datastoreAdapter.storeResult(buildTask, buildResult);
+                coordinationStatus = BuildCoordinationStatus.DONE;
+            }
+        } catch (DatastoreException | BuildCoordinationException e ) {
+            log.error("Cannot store results to datastore.", e);
+            coordinationStatus = BuildCoordinationStatus.SYSTEM_ERROR;
+        } finally {
+            buildTask.setStatus(coordinationStatus);
+            markFinished(buildTask);
+        }
     }
 
-    public List<BuildTask> getSubmittedBuildTasks() {
-        return Collections.unmodifiableList(submittedBuildTasks.stream().collect(Collectors.toList()));
+    public synchronized void markFinished(BuildTask task) {
+        buildQueue.removeTask(task);
+        switch (task.getStatus()) {
+            case DONE:
+            case REJECTED_ALREADY_BUILT:
+                buildQueue.executeNewReadyTasks();
+                break;
+            case REJECTED:
+            case REJECTED_FAILED_DEPENDENCIES:
+            case SYSTEM_ERROR:
+            case DONE_WITH_ERRORS:
+                handleErroneousFinish(task);
+                break;
+            default:
+                throw new IllegalArgumentException("Unhandled build task status: " + task.getStatus() + ". Build task: " + task);
+        }
+
+        BuildSetTask buildSetTask = task.getBuildSetTask();
+        if (buildSetTask != null && isFinished(buildSetTask)) {
+            completeBuildSetTask(buildSetTask);
+        }
+    }
+
+    private boolean isFinished(BuildSetTask buildSetTask) {
+        return buildSetTask
+                .getBuildTasks()
+                .stream()
+                .allMatch(t -> t.getStatus().isCompleted());
     }
 
-    @Deprecated //Used only in tests
-    public boolean hasActiveTasks() {
-        if (log.isTraceEnabled()) {
-            String activeTasks = submittedBuildTasks.stream().map(bt -> bt.getId() + "-" + bt.getStatus()).collect(Collectors.joining(","));
-            log.trace("Build Coordinator Active Tasks {}", activeTasks);
+    private void handleErroneousFinish(BuildTask failedTask) {
+        BuildSetTask taskSet = failedTask.getBuildSetTask();
+        if (taskSet != null) {
+            taskSet.getBuildTasks().stream()
+                    .filter(t -> isDependentOn(failedTask, t))
+                    .forEach(t -> finishDueToFailedDependency(failedTask, t));
         }
-        return submittedBuildTasks.peek() != null;
     }
 
-    private boolean isBuildAlreadySubmitted(BuildTask buildTask) {
-        return submittedBuildTasks.contains(buildTask);
+    private boolean isDependentOn(BuildTask failedTask, BuildTask t) {
```

Comment:
Could you add some api doc and maybe use some more descriptive parameter name instead of `t`.